packages/expo-video/ios/Advertisement/VideoAdsManagerIMA.swift

```diff
@@ -0,0 +1,207 @@
+          fullscreenController.present(adVideoView, animated: false, completion: nil)
+      }
+    }
+    
+    // Exit fullscreen mode
+    public func exitFullscreen() {
+      fullscreenController?.dismiss(animated: false) {
+        self.fullscreenController = nil
+      }
+    }
+     
+    // Create an ad request with our ad tag,
+    public func requestAds(adDisplayContainer: AdDisplayContainer, adTagUri: String){
+        self.adDisplayContainer = adDisplayContainer
+        
+        reset()
+        let request = IMAAdsRequest(
+            adTagUrl: adTagUri,
+            adDisplayContainer: adDisplayContainer,
+            contentPlayhead: contentPlayhead,
+            userContext: nil
+        )
+
+        adsLoader.requestAds(with: request)
+    }
+        
+    // Triggered when normal video has finished
+    func contentDidFinishPlaying() {
+      // Show post-roll if requested.
+      isContentCompleteCalled = true
+      adsLoader.contentComplete()
+    }
+    
+    func prepareAds(player: AVPlayer, videoPlayerItem: VideoPlayerItem?, videoView: VideoView?){
+      let advertisement = videoPlayerItem?.videoSource.advertisement?.googleIMA?.adTagUri
+          
+      if let adTagUri = advertisement, let videoView = videoView {
+        // TODO: Set loading ads (waitForPreroll config?)
+        let adDisplayContainer = AdDisplayContainer( adContainer: videoView.playerViewController.view,  viewController: videoView.playerViewController)
+              
+        requestAds(
+          adDisplayContainer: adDisplayContainer,
+          adTagUri: adTagUri
+        )
+      }
+    }
+    
+    // MARK: - IMAAdsLoaderDelegate
+    
+    func adsLoader(_ loader: IMAAdsLoader, adsLoadedWith adsLoadedData: IMAAdsLoadedData) {
+      adsManager = adsLoadedData.adsManager
+      adsManager?.delegate = self
+      adsManager?.initialize(with: nil)
+      print("AdLoader successfully initialized")
+    }
+
+    func adsLoader(_ loader: IMAAdsLoader, failedWith adErrorData: IMAAdLoadingErrorData) {
+        print("Error loading ads: " + (adErrorData.adError.message ?? "No message"))
+        self.isPlayingAd = false
+    }
+    
+    // MARK: - IMAAdsManagerDelegate
+
+    public func adsManager(_ adsManager: IMAAdsManager, didReceive event: IMAAdEvent) {
+        print("AdsManager Event: " + event.typeString)
+        
+        switch event.type {
+          case IMAAdEventType.LOADED:
+            // Queue ads when they are ready
+            hasMoreAds = true
+            adsManager.start()
+            break
+          case IMAAdEventType.STARTED:
+            // Set Ads to fullscreen if the AVPlayer is in fullscreen
+            if isContentFullscreen {
+              enterFullscreen()
+            }
+            break
+          case IMAAdEventType.ALL_ADS_COMPLETED:
+            hasMoreAds = false
+            
+            // Trigger video complete if video has finished
+            if(isContentCompleteCalled) {
```

Comment:
Nit: keep an eye on these formatting issues. Are you using VS Code instead of Xcode to write Swift?